Products/zms/zms.py

```diff
@@ -1,606 +1,639 @@
+
+def subscriber(event):
+  if isinstance(event, ObjectAddedEvent):
+    if isinstance(event.object, ZMSObject):
+      if isinstance(event.newParent, ZMSObject):
+        # trigger object-added event
+        standard.triggerEvent(event.object, "*.ObjectAdded")
+  elif isinstance(event, ObjectMovedEvent):
+    if isinstance(event.object, ZMSObject):
+      if isinstance(event.newParent, ZMSObject):
+        # trigger object-moved event
+        standard.triggerEvent(event.object, "*.ObjectMoved")
+      elif event.newParent is None:
+        standard.triggerEvent(event.object, "*.ObjectRemoved")
+  elif isinstance(event, ObjectRemovedEvent):
+    if isinstance(event.object, ZMSObject):
+      # trigger object-removed event
+      standard.triggerEvent(event.object, "*.ObjectRemoved")
+zope.event.subscribers.append(subscriber)
+
+
+################################################################################
+#
+# Common Function(s)
+#
+################################################################################
+
+# ------------------------------------------------------------------------------
+#  importTheme:
+# ------------------------------------------------------------------------------
+def importTheme(self, theme):
+  filename = _fileutil.extractFilename(theme)
+  id = filename[:filename.rfind('-')]
+  filepath = package_home(globals()) + '/import/'
+  path = filepath + filename
+  self.importConf(path)
+  return id
+
+
+# ------------------------------------------------------------------------------
+#  initZMS:
+# ------------------------------------------------------------------------------
+# A new ZMS node can be initalized as a stand-alone client (master) or
+# as subordinated client acquiring content models and sharing the zmsindex.
+# Use a request variable 'acquire' =  1 to initalize ZMS as a client
+def initZMS(self, id, titlealt, title, lang, manage_lang, REQUEST):
+
+  ### Constructor.
+  obj = ZMS()
+  obj.id = id
+  self._setObject(obj.id, obj)
+  obj = getattr(self, obj.id)
+
+  ### Trashcan.
+  trashcan = ZMSTrashcan()
+  obj._setObject(trashcan.id, trashcan)
+
+  ### Manager.
+  manager = ZMSMetamodelProvider.ZMSMetamodelProvider()
+  obj._setObject( manager.id, manager)
+  manager = ZMSMetacmdProvider.ZMSMetacmdProvider()
+  obj._setObject( manager.id, manager)
+  manager = ZMSFormatProvider.ZMSFormatProvider()
+  obj._setObject( manager.id, manager)
+
+  ### Init languages.
+  obj.setLanguage(lang, REQUEST['lang_label'], '', manage_lang)
+
+  ### Log.
+  if REQUEST.get('zmslog'):
+    zmslog = ZMSLog( copy_to_stdout=True, logged_entries=[ 'ERROR', 'INFO'])
+    obj._setObject(zmslog.id, zmslog)
+
+  ### Init Configuration.
+  obj.setConfProperty('HTTP.proxy', REQUEST.get('http_proxy', ''))
+  obj.setConfProperty('ZMS.autocommit', 1)
+
+  ### Init ZMS default content-model.
+  _confmanager.initConf(obj, 'conf:com.zms.foundation')
+  _confmanager.initConf(obj, 'conf:com.zms.catalog.zcatalog')
```

Comment:
@drfho likes it
should only by imported when ZMI inserted - not via API call etc.

Reply:
Idea:

```suggestion
  _confmanager.initConf(obj, partial and 'conf:com.zms.foundation' or 'conf:com.zms.foundation*')
```

```python
def initZMS(self, id, titlealt, title, lang, manage_lang, REQUEST, partial = False):
```

Reply:
We still have to analyse if we really need any content-model inits.

Reply:
DONE by https://github.com/zms-publishing/ZMS/commit/b7ad7d72fa141b6f098ccda789c00d7e6880a95f

Reply:
and https://github.com/zms-publishing/ZMS/commit/a637c49799f228daf973f8860473c67337dfd280